surface/include/pcl/surface/impl/ball_pivoting.hpp

```diff
@@ -0,0 +1,629 @@
+  
+  /**
+   * get the center of circle where three point are on
+   * @param point0
+   * @param point1
+   * @param point2
+   * @return
+   */
+  Eigen::Vector3f
+  get_circle_center (const Eigen::Vector3f &point0, const Eigen::Vector3f &point1, 
+                     const Eigen::Vector3f &point2)
+  {
+    // https://en.wikipedia.org/wiki/Circumscribed_circle#Cartesian_coordinates_from_cross-_and_dot-products
+    const Eigen::Vector3f vec2 = point0 - point1;
+    const Eigen::Vector3f vec0 = point1 - point2;
+    const Eigen::Vector3f vec1 = point2 - point0;
+    const float area = vec0.cross (vec1).norm ();
+    const float determinator = 2.0f * area * area;
+  
+    const float alpha = vec0.dot (vec0) * vec2.dot (-vec1) / determinator;
+    const float beta = vec1.dot (vec1) * -vec2.dot (vec0) / determinator;
+    const float gamma = vec2.dot (vec2) * vec1.dot (-vec0) / determinator;
+  
+    return alpha * point0 + beta * point1 + gamma * point2;
+  }
+  
+  /**
+   * get the normal vector of a triangle, (p1-p0)x(p2-p0)
+   * @tparam PointNT
+   * @param cloud
+   * @param index
+   * @return
+   */
+  template<typename PointNT>
+  Eigen::Vector3f
+  get_normal_triangle (const typename pcl::PointCloud<PointNT>::ConstPtr &cloud, 
+                       const std::vector<uint32_t> &indexes)
+  {
+    if (indexes.size () != 3)
+    {
+      PCL_ERROR ("indexes.size () != 3 in get_normal_triangle");
+      return Eigen::Vector3f::Zero ();
+    }
+    const Eigen::Vector3f p0 = cloud->at (indexes.at (0)).getVector3fMap ();
+    return (cloud->at (indexes.at (1)).getVector3fMap () - p0)
+      .cross (cloud->at (indexes.at (2)).getVector3fMap () - p0)
+      .normalized ();
+  }
+  
+  /**
+   * reorder the vertices of triangle so the normal vector of triangle is consistent with normals of vertices
+   * @tparam PointNT
+   * @param cloud
+   * @param triangle
+   */
+  template<typename PointNT>
+  void
+  reorder (const typename pcl::PointCloud<PointNT>::ConstPtr &cloud, pcl::Vertices &triangle)
+  {
+    if (!is_normal_consistent (get_normal_triangle<PointNT> (cloud, triangle.vertices), 
+                               triangle.vertices, cloud))
+    {
+      // if the order (0,1,2) is not consistent, (0,2,1) should be consistent
+      std::swap (triangle.vertices.at (1), triangle.vertices.at (2));
+    }
+  }
+  
+  /**
+   * get the distance from point to plane
+   * @tparam PointNT
+   * @param plane
+   * @param point
+   * @return
+   */
+  template<typename PointNT>
+  float
+  get_distance_point_plane (const Eigen::Vector4f &plane, const PointNT &point)
+  {
+    // plane is (a,b,c,d), point is (x,y,z), then distance is ax+by+cz+d
+    return point.getVector3fMap ().dot (plane.segment (0, 3)) + plane (3);
+  }
+  
+  /**
+   * get the signed rotation angle from (point0-center) to (point1-center) on plane
+   * @param point0
+   * @param point1
+   * @param center
+   * @param plane the rotation is along the normal vector of plane
```

Comment:
The direction can be inferred in normal vector of the plane, either alpha in one direction or 360-alpha in the other